### DISCUSSION TITLE
Errors after upgrading to 9.8

### DISCUSSION THREAD

**Discussion by allanedk** [dnnsoftware/Dnn.Platform#4402] [General] [answered]
After upgrading from DNN 9.7.2 to DNN 9.8 (without issues) I get errors like this:

```
The type 'System.Collections.Generic.ICollection`1<T0>' is defined in an assembly that is not referenced. Add a reference to the assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
```

I tried changed the code that caused the error, but realized that variations of the error occurs in a number of different situations like "The type 'Enum' (...)" etc. 

When I manually add this reference to the web.config everything works fine. But I don't think that adding it manually is the right solution.

```
<add assembly="netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" />
```

Any idea what could be causing this?

I have verified that .NET Framework 4.7.2 is installed.

**Comment by bdukes** [comment]
I assume there's a `netstandard.dll` file in the `bin` folder, yes?  This doesn't come with DNN, so you'll want to determine which extension you're using added that file.  That component should probably be updated to add the `<add assembly="netstandard, …"/>` entry to the `web.config`, or not include the file if it's not needed.

**Reply by allanedk** [reply]
Thanks a lot for responding. 

I checked and there is no netstandard.dll in the bin dir.

Could this be an issue with 9.8?

I tried to do the upgrade several times now with the same result. I'm only using a few core modules and some custom developed modules. No code was changed in my own modules other than upgrading Nuget packages to DNN9.8 and rebuild.

The error occurs in this code, which worked fine in DNN9.7:

```
var objEventLog = new EventLogController();
objEventLog.AddLog("text", "text", PortalSettings, UserId, EventLogController.EventLogType.ADMIN_ALERT);
```

And when I change that to use IEventLogger instead, a similar error occurs saying that "Enum" (with a capital E, not "enum") is defined in an assembly not referenced (netstandard).

```
[...]
eventLogger.AddLog("text", "text", null, null, UserId, DotNetNuke.Abstractions.Logging.EventLogType.ADMIN_ALERT);
```

**Comment by bdukes** [comment]
I don't think I've seen that issue before.  It _should_ be safe to add that `<add assembly="netstandard, …"/>` entry into the `web.config`.

To track down the ultimate issue, it would be worth comparing the assemblies included in the install package when you're compiling against DNN 9.7.x and DNN 9.8.x, and :one: make sure you're not including assemblies that DNN provides, and :two: make sure there aren't any surprises about new or changed assemblies (in theory, only your custom assemblies would change).

**Comment by allanedk** [comment]
Okay: I have spend **a lot** of time now doing code cleanup followed by step by step upgrading an comparing changes (as suggested). I found nothing odd at all.

As a sanity check - which I probably should have done as a first step :) - I created a simple module in a fresh DNN 9.8 installation. It only adds a log entry to the DNN event log and nothing else. It's basically just a ASCX file registered as a module. It doesn't work either. This is the source code.

```
using System;
using System.Collections.Generic;
using DotNetNuke.Abstractions.Logging;
using DotNetNuke.Services.Log.EventLog;

public partial class DesktopModules_Test : DotNetNuke.Entities.Modules.PortalModuleBase
{

    protected void Page_Load(object sender, EventArgs e)
    {
        var objEventLog = new EventLogController();
        objEventLog.AddLog("Log using old EventLogController", "", PortalSettings, UserId, EventLogController.EventLogType.ADMIN_ALERT);
    }
   
}

```
The same code worked in DNN9.7 and works when I manually add a reference to netstandard in web.config. 

I realized that this might only to be an issue in webforms module. The same code works in my webservices (DnnApiController).

Do you think this is an DNN98 issue or am I missing something here?


PS: I know that the EventLogController is deprecated in 9.7.

**Reply by bdukes** [reply]
Can you upload the package to your test module, so we can review if there's anything out of the ordinary, or if we can replicate?  Thanks for putting in the effort to help track this down 😄

**Comment by allanedk** [comment]
Sure thing. Thanks for helping me out.

There no install package, but it can be registered manually:

1. Unzip and place folder in DesktopFolders
2. Go to Admin -> Extensions
3. Click "Add new module"
4. Choose "Create new module from: Control"
5. Select module folder: "TestMissingAssembly" and create
6. Add module a page
7. When loaded you should see either a Host alert entry in the event log or an error warning about the missing reference to netstandard.

[TestMissingAssembly.zip](https://github.com/dnnsoftware/Dnn.Platform/files/5787886/TestMissingAssembly.zip)

**Reply by bdukes** [reply]
I am definitely seeing the same issue with this control.  Are you using this same setup for your real modules, where the `.ascx.cs` file is included, rather than being compiled?  I'm trying to understand if it's an issue with the ASP.NET compilation process, or if it also affects `dll` files.

**Reply by bdukes** [reply]
To resolve the issue with the test control, I opened the _Extensions_ page in the Persona Bar, selected the _Available Extensions_ tab, selected _Libraries_ from the _Showing:_ drop down list, and installed _Microsoft CodeDom .NET Compiler for .NET 4.6+_.  This upgrades the compiler used by ASP.NET.  I assume this adds support for .NET Standard during that compilation process.  I'm not, however, sure why this became necessary during the 9.8.0 upgrade.  Unfortunately, we can't automatically install this package as part of DNN Platform because it includes `csc.exe` (the compiler executable) and some hosting providers don't allow deploying `exe` files.

**Reply by bdukes** [reply]
Interesting, the issue is still there when using `IEventLogger` (with the `Enum` is not referenced message)…

**Comment by allanedk** [comment]
Thank you for looking into this. Do you see installing the Microsoft CodeDom .NET Compiler as a solution or as a temporary workaround? In other words: Can I install it and upgrade my production site to 9.8?

And yes, I use this setup with (fairly simple) uncompiled modules (including .ascx.cs) in combination with a more complex (compiled) WebAPI.

**Comment by allanedk** [comment]
Ah, I see that you added another reply after I started writing mine. Seems that it's not a solid, final solution :)

**Comment by bdukes** [comment, marked as answer]
Using the CodeDOM provider will allow you to use newer compiler features but isn't a complete fix.  I think you will need to go ahead and add the `netstandard` reference if you're not going to compile your codebehind classes.  You can add [a config component](https://dnndocs.com/content/tutorials/extensions/dnn-manifest-schema/index.html#components) to your manifest using the following [XML Merge](https://www.dnnsoftware.com/wiki/manifest-xml-merge) document:

```xml
<configuration>
  <nodes configfile="Web.config">
    <node path="/configuration/system.web/compilation/assemblies"
          action="update"
          key="assembly"
          collision="ignore">
      <add assembly="netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" />
    </node>
  </nodes>
</configuration>
```

You can also run this manually in the _Config Manager_ page in the Persona Bar.

**Reply by allanedk** [reply]
Okay, thanks a lot for the help, Brian.

I'll go on and add the reference to web.config for now (it works like a charm in my dev-environment). Later I will probably need to change the setup to use pre-compiled modules or - even better - Razor pages modules if/when it becomes an option in DNN.
